documentation/src/main/asciidoc/topics/query.adoc

```diff
@@ -18,480 +18,570 @@ Apart from indexed queries, {brandname} can run queries over non-indexed or
-letting {brandname} link:#query_autoconfig[configure the indexing automatically]:
+=== Configuration
 
-{brandname} configuration:
+To enable indexing via XML, you need to add the `<indexing>` element to your cache configuration, specify the entities that are indexed and optionally pass additional properties.
+
+NOTE: The presence of an `<indexing>` element which omits the `enabled` attribute will auto-enable indexing for your
+convenience, even though the default value of the `enabled` attribute is defined as `"false"` in the XSD schema.
+This behaviour exists only on schemas starting with version 11.
 
-.infinispan.xml
 [source,xml,options="nowrap",subs=attributes+]
 ----
-include::config_examples/infinispan_distributed_cache_books.xml[]
+include::config_examples/indexing.xml[]
 ----
 
-Obtaining the cache:
+Programmatic:
 
 [source,java]
 ----
-import org.infinispan.Cache;
-import org.infinispan.manager.DefaultCacheManager;
-import org.infinispan.manager.EmbeddedCacheManager;
-
-EmbeddedCacheManager manager = new DefaultCacheManager("infinispan.xml");
-Cache<String, Book> cache = manager.getCache("books");
+import org.infinispan.configuration.cache.*;
 
+ConfigurationBuilder cacheCfg = ...
+cacheCfg.indexing().enable()
+            .addIndexedEntity(Book.class)
+      .addProperty("property name", "propery value")
 ----
 
-Each _Book_ will be defined as in the following example; we have to choose which properties are indexed, and for each property we can optionally choose advanced indexing options using the annotations defined in the Hibernate Search project.
-
-[source,java]
-.Book.java
-----
-import org.hibernate.search.annotations.*;
-import java.util.Date;
-import java.util.HashSet;
-import java.util.Set;
+NOTE: When using programmatic configuration, indexing must be enabled explicitly by invoking the `enable()` method, as
+no auto-enabling mechanism exist for programmatic configuration as in the case of XML configuration.
 
-//Values you want to index need to be annotated with @Indexed, then you pick which fields and how they are to be indexed:
-@Indexed
-public class Book {
-   @Field String title;
-   @Field String description;
-   @Field @DateBridge(resolution=Resolution.YEAR) Date publicationYear;
-   @IndexedEmbedded Set<Author> authors = new HashSet<Author>();
-}
+=== Specifying indexed Entities
 
-----
+It is recommended to declare the indexed types, as they will be mandatory in the next {brandname} version.
+This can be done via xml:
 
-[source,java]
-.Author.java
+[source,xml,options="nowrap",subs=attributes+]
 ----
-
-public class Author {
-   @Field String name;
-   @Field String surname;
-   // hashCode() and equals() omitted
-}
-
+include::config_examples/indexed_entities.xml[]
 ----
 
-Now assuming we stored several _Book_ instances in our {brandname} _Cache_ , we can search them for any matching field as in the following example.
-
-Using a Lucene Query:
+or programmatically:
```

Comment:
One thing we've been doing elsewhere in docs, like marshalling, is to separate declarative and programmatic examples like this:

.Declaratively
xml snippet goes here

.Programmatically
code snippet goes here